joint_state_controller:
  type: joint_state_controller/JointStateController
  publish_rate: 50

diff_drive_controller:
  type: diff_drive_controller/DiffDriveController
  left_wheel: ['front_left_wheel_joint', 'rear_left_wheel_joint']
  right_wheel: ['front_right_wheel_joint', 'rear_right_wheel_joint']
  publish_rate: 10
  wheel_separation: 0.4
  wheel_radius: 0.08
  base_frame_id: base_footprint
  cmd_vel_timeout: 0.25
  enable_odom_tf: true

velocity_controller:
  type: velocity_controllers/JointGroupVelocityController
  joints:
    - front_left_wheel_joint
    - front_right_wheel_joint
    - rear_left_wheel_joint
    - rear_right_wheel_joint

# 前轮转向控制器（ackermann_front用）
front_left_steering_controller:
  type: position_controllers/JointPositionController
  joint: front_left_steering_joint

front_right_steering_controller:
  type: position_controllers/JointPositionController
  joint: front_right_steering_joint

# 后轮转向控制器（ackermann_rear用）
rear_left_steering_controller:
  type: position_controllers/JointPositionController
  joint: rear_left_steering_joint

rear_right_steering_controller:
  type: position_controllers/JointPositionController
  joint: rear_right_steering_joint

#gazebo_ros_control:
#  pid_gains:
#    front_left_wheel_joint:
#      p: 2.0    # 比例增益，必须有
#      d: 1.0      # 微分增益，可调
#    front_right_wheel_joint:
#      p: 2.0    # 比例增益，必须有
#      d: 1.0      # 微分增益，可调
#    rear_left_wheel_joint:
#      p: 2.0    # 比例增益，必须有
#      d: 1.0      # 微分增益，可调
#    rear_right_wheel_joint:
#      p: 2.0    # 比例增益，必须有
#      d: 1.0      # 微分增益，可调
#    front_right_wheel_joint: {p: 100.0, i: 0.1, d: 1.0, i_clamp: 1.0}
#    rear_left_wheel_joint: {p: 100.0, i: 0.1, d: 1.0, i_clamp: 1.0}
#    rear_right_wheel_joint: {p: 100.0, i: 0.1, d: 1.0, i_clamp: 1.0}
#    front_left_steering_joint: {p: 100.0, i: 0.1, d: 1.0, i_clamp: 1.0}
#    front_right_steering_joint: {p: 100.0, i: 0.1, d: 1.0, i_clamp: 1.0}
#    rear_left_steering_joint: {p: 100.0, i: 0.1, d: 1.0, i_clamp: 1.0}
#    rear_right_steering_joint: {p: 100.0, i: 0.1, d: 1.0, i_clamp: 1.0}